### DISCUSSION TITLE
Fetching data off the main thread

### DISCUSSION THREAD

**Discussion by mark-looptimize** [lit/lit#2221] [Q&A]
I was having a bit of trouble understanding the best way to approach data fetching from a performance point of view with Lit.

I identified the following native and Lit primitives which sound like they could be helpful.

- Workers (to do the heavy work in another thread)
- ReactiveControllers allow me to keep the code clean from my component
- Lit's Task API

However, I am not sure I ended up in a good place and wanted to get some feedback on this. What I had in mind was:

1. Set up a ReactiveController and host it inside of a component that needs data
2. Use the controllers life cycle hook to call the Task API
3. Task API actually sends a message to a Worker 
4. Worker fetches and returns the data
5. Controller updates the component

That seems "technically correct" and also like a lot so I wanted to get some advice on this please :)

**Comment by nemethmik** [comment]
Unless you insist on using *@lit-lab/task* here is a regular solution.
```ts
type TPerson = {
  firstName: string,
  lastName: string,
  email: string
}
@customElement("boot-people")
class BootPeople extends BootBase {
  @state() people:TPerson[] = []
  override connectedCallback():void {
    super.connectedCallback()
    this.loadData()
  }
  async loadData():Promise<void> {
    //Call this from the constructor. Don't call it from firstUpdate
    //THIS IS VERY IMPORTANT TO GIVE a chance to the component to construct itself
    await new Promise((r) => setTimeout(r, 0))
    await new Promise((r) => setTimeout(r, 1000)) // This is just a delayer to simulate long running
    const endpoint = "./people.json"
    this.people = (await (await fetch(endpoint)).json()).result
  }
  override render():TemplateResult {return html`
    <p>
    ${this.people.length ? html`<h2>Number of people ${this.people.length}</h2>` : html`
      <div class="spinner-grow text-primary" role="status">
        <span class="visually-hidden">Loading...</span>
      </div>
    `}
    </p>
  `}  
}
declare global { interface HTMLElementTagNameMap { "boot-people": BootPeople}}
```
Here I use Bootstrap spinner defined in my base class BootBase.
With lit-mobx it is even simpler, let me know if you are interested in a sample.

**Reply by mark-looptimize** [reply]
Thanks for the reply on this. The issue I had here was that this didn't use the _@labs/task_ or a worker here to move that off the main thread. What you presented though is helpful as a nice example for one way to do things that was really clear and simple which I genuinely appreciate! 

However, I am also interested in what the trade off looks like for trying to bring in those choices I mentioned and even understand if they are even able to work together or not.
